### PR TITLE
[Merged by Bors] - Support multiple signers in beacon

### LINKED ISSUE #5088: beacon: support multiple smeshers
please see https://github.com/spacemeshos/go-spacemesh/issues/5085 for:
- consistent interface
- parallelization considerations
- validation on self-publishing optimization

one another bottleneck that we need to be careful about is not to emit all messages at once. basically [jitter](https://github.com/spacemeshos/go-spacemesh/issues/4989) needs to be considered for every identity separately. one working option to implement it would be to use time.AfterFunc

### PR DESCRIPTION
**Pull request by poszu**
## Motivation
Part of https://github.com/spacemeshos/pm/issues/261
Closes #5088

## Changes
- beacon protocol driver supports registering new keys to participate in the protocol
- when a new epoch starts, the protocol selects active signers that will participate
- state is changed to contain the set of active protocol participants. A participant session consists of:
  * signer
  * VRF nonce
- weakcoin supports multiple participants. Each one publishes the same proposal.

## Test Plan
Existing unit-tests were extended to run beacon and weakcoin with multiple signers per node.

### CODE REVIEW

File: beacon/beacon.go [hunk 1]
```diff
@@ -135,22 +120,60 @@ func New(
...
-		pd.nonceFetcher = defaultFetcher{cdb: cdb}
-	}
 
-	if pd.weakCoin == nil {
-		pd.weakCoin = weakcoin.New(pd.publisher, edSigner.VRFSigner(), vrfVerifier, pd.nonceFetcher, pd,
-			pd.msgTimes,
-			weakcoin.WithLog(pd.logger.WithName("weakCoin")),
-			weakcoin.WithMaxRound(pd.config.RoundsNumber),
-		)
+	if pd.createWeakCoin == nil {
```
Comment (dshulyak) [inline beacon/beacon.go]: it doesn't make sense to have several weak coin instances. there should be on instance that holds state, and routine that goes over all signers and creates proposal for each one

in current implementation i think the simplest would be to split StartRound into two methods, one generic for all signers and another one accepts signer and publishes proposal

File: beacon/beacon.go [hunk 2]
```diff
@@ -848,21 +879,42 @@ func (pd *ProtocolDriver) runConsensusPhase(ctx context.Context, epoch types.Epo
...
+
+	// First round
+	round := types.FirstRound
+	pd.setRoundInProgress(round)
+	if msg, err := pd.genFirstRoundMsgBody(epoch); err != nil {
+		logger.With().Error("failed to get firstround message", log.Err(err), round)
+	} else {
+		for _, session := range active {
+			session := session
+			pd.eg.Go(func() error {
```
Comment (dshulyak) [inline beacon/beacon.go]: could you remove pb.eg.Go in sendToGossip?

File: beacon/handlers.go [hunk 3]
```diff
@@ -38,12 +40,21 @@ var (
...
 	}
+	var eg errgroup.Group
+	pd.mu.RLock()
+	for _, signer := range pd.signers {
+		coin := signer.coin
+		eg.Go(func() error {
+			return coin.HandleProposal(ctx, peer, msg)
+		})
+	}
+	pd.mu.Unlock()
```
Comment (dshulyak) [inline beacon/handlers.go]: should RUnlock? but i think this can be refactored to avoid creating multiple weak coin instances

File: beacon/handlers.go [hunk 4]
```diff
@@ -229,17 +240,17 @@ func (pd *ProtocolDriver) HandleFirstVotes(ctx context.Context, peer p2p.Peer, m
 	currentEpoch := pd.currentEpoch()
 	if m.EpochID != currentEpoch {
 		logger.With().Debug("first votes from different epoch",
-			log.Uint32("current_epoch", uint32(currentEpoch)),
-			log.Uint32("message_epoch", uint32(m.EpochID)))
+			log.FieldNamed("current_epoch", currentEpoch),
+			log.FieldNamed("message_epoch", m.EpochID))
```
Comment (dshulyak) [inline beacon/handlers.go]: FieldNamed is deprecated as it relies on copying whole array and then unpacking original fields

File: beacon/handlers.go [hunk 5]
```diff
@@ -229,17 +240,17 @@ func (pd *ProtocolDriver) HandleFirstVotes(ctx context.Context, peer p2p.Peer, m
...
 	}
 
 	// don't accept more first vote after the round ends
 	currentRound := pd.currentRound()
 	if currentRound > types.FirstRound {
 		logger.With().Debug("first votes too late",
-			log.Uint32("current_round", uint32(currentRound)),
-			log.Uint32("message_round", uint32(types.FirstRound)))
+			log.FieldNamed("current_round", currentRound),
+			log.FieldNamed("message_round", types.FirstRound))
```
Comment (dshulyak) [inline beacon/handlers.go]: FieldNamed is deprecated

File: beacon/beacon.go [hunk 6]
```diff
@@ -135,22 +120,60 @@ func New(
...
+
+func (s *participant) Id() signerID {
+	return signerID(s.signer.NodeID())
+}
+
+type signerID types.NodeID
+
+func (id signerID) Field() log.Field {
+	return log.ShortStringer("id", types.NodeID(id))
+}
```
Comment (fasmat) [inline beacon/beacon.go]: I'm not sure about introducing a new type here just to change the logging behavour of `NodeID` only for `beacon`. We should consider `(NodeID) Field()` to return a `log.ShortStringer` and have that behaviour everywhere.

File: beacon/beacon.go [hunk 7]
```diff
@@ -848,21 +879,42 @@ func (pd *ProtocolDriver) runConsensusPhase(ctx context.Context, epoch types.Epo
...
 		err       error
 	)
-	for round := types.FirstRound; round < pd.config.RoundsNumber; round++ {
+
+	// First round
+	round := types.FirstRound
+	pd.setRoundInProgress(round)
+	if msg, err := pd.genFirstRoundMsgBody(epoch); err != nil {
+		logger.With().Error("failed to get firstround message", log.Err(err), round)
+	} else {
```
Comment (fasmat) [inline beacon/beacon.go]: The only reason `pd.genFirstRoundMsgBody` might fail is if the identity isn't active in the given epoch. But `active` already has been checked for that so can this method return an error?

Alternatively `pd.genFirstRoundMsgBody` could just be part of `sendFirstRoundVote` to simplify the logic.

File: beacon/handlers.go (hunk 4 above)
Comment (poszu) [reply beacon/handlers.go]: I don't understand - what array do you have in mind and where is it unpacked? Is the problem that `func (e EpochID) Field()` is called eagerly inside of `FieldNamed()`?

Anyway - if it's deprecated let's mark it so and create an issue to delete it.

File: beacon/beacon.go (hunk 6 above)
Comment (poszu) [reply beacon/beacon.go]: What's wrong with having a private type here? Btw, it changed the key as well (NodeID logs "node_id", not "id").

Comment (fasmat) [reply beacon/beacon.go]: The problem isn't the private type, it's that `beacon` now logs `NodeID` differently from all other components, which makes it harder to read them imo.

File: beacon/beacon.go (hunk 7 above)
Comment (poszu) [reply beacon/beacon.go]: It's generally a problem with creating too many "small helper methods" that it is hard to follow in what conditions a method is called (for example, what is already asserted). I looked over code again. It turns out `genFirstRoundMsgBody()` is called in 1 place and the `state` for the epoch has already been picked and asserted that it exist. Hence it makes no sense to pick it up again from `pd.states[epoch]` in the first place.

I will refactor the code to pass the once picked state down the stack.

>Alternatively pd.genFirstRoundMsgBody could just be part of sendFirstRoundVote to simplify the logic.

It cannot because:
- this work is common for all signers so it makes no sense to repeat it
- it mutates shared state by sorting incoming proposals
 
That's why I extracted it from `sendFirstRoundVote`.

Comment (fasmat) [reply beacon/beacon.go]: > this work is common for all signers so it makes no sense to repeat it

I initially assumed the method just creates a message and returns -> that doesn't need to be its own method. The problem here is that the method has an additional side effect (sorting proposals) that isn't not obvious from its name or on a first glance when looking at the method.

> it mutates shared state by sorting incoming proposals

That's fair, but then sorting should be done once in a method that's named "sortIncomingProposals" or similar and the instantiation of the Message object can be done in `sendFirstRoundVote`. At least I think that would make it clearer what's happening 🙂

File: beacon/beacon.go (hunk 6 above)
Comment (poszu) [reply beacon/beacon.go]: IMHO, it doesn't make sense to log messages per-signer with a "node_id" field, it has nothing to do with "node" anymore (it's not a node ID in this context, there are many IDs governed by a single node).

Perhaps instead, the `NodeID` type should be dropped entirely in favor of a `SignerID` or something similar? But that's a bigger refactor.

File: beacon/beacon.go (hunk 7 above)
Comment (poszu) [reply beacon/beacon.go]: I removed the method entirely. ~Btw, I discovered that it had a bug - it took a shared Rlock and mutated the data :(~

File: beacon/handlers.go (hunk 3 above)
Comment (poszu) [reply beacon/handlers.go]: Good catch, thanks! The sad part is that it means this code is never executed in tests :(

File: beacon/handlers.go (hunk 4 above)
Comment (dshulyak) [reply beacon/handlers.go]: i meant this code in spacemesh logger

```go
func (fl FieldLogger) Info(msg string, fields ...LoggableField) {
	fl.l.Info(msg, unpack(append(fields, String("name", fl.name)))...)
}
```

File: beacon/beacon.go (hunk 1 above)
Comment (poszu) [reply beacon/beacon.go]: You are right, I refactored the code, but I didn't split the `StartRound` method as it didn't seem necessary. It now takes a slice of participants (vrf signer + vrf nonce).

File: beacon/handlers.go (hunk 4 above)
Comment (poszu) [reply beacon/handlers.go]: OK, but this happens anyway, doesn't it?

**Comment by poszu** [comment]
bors try

File: beacon/weakcoin/weak_coin.go [hunk 8]
```diff
@@ -193,7 +192,7 @@ func (wc *WeakCoin) FinishEpoch(ctx context.Context, epoch types.EpochID) {
 	if epoch != wc.epoch {
 		logger.With().Fatal("attempted to finish beacon weak coin for the wrong epoch",
 			epoch,
-			log.Stringer("weak_coin_epoch", wc.epoch),
+			log.FieldNamed("weak_coin_epoch", wc.epoch),
```
Comment (dshulyak) [inline beacon/weakcoin/weak_coin.go]: can you Uint32(), my idea was that we remove FieldNamed over time and then it reduces amount of refactoring

**Review by dshulyak — APPROVED** [review]

File: beacon/handlers.go (hunk 4 above)
Comment (fasmat) [reply beacon/handlers.go]: Would it be much effort to migrate `ProtocolDriver` from using `go-spacemesh/log` to using `zap`? I also don't fully understand why loggers are passed around via context, when `ProtocolDriver` already has an instance of a logger?

**Comment by poszu** [comment]
bors try

File: beacon/handlers.go (hunk 4 above)
Comment (poszu) [reply beacon/handlers.go]: Probably not much effort. I will try to migrate this/next week.

**Comment by poszu** [comment]
bors merge

**Comment by poszu** [comment]
Failure in unrelated UT

bors merge
